### ISSUE TITLE
Inaccurate output when converting bool array to bfloat16 

### ISSUE THREAD

**Issue by bzhangGo** [google/jax#18387]
### Description

```python
import jax
import numpy as np

np.random.seed(42)

x = (np.random.rand(2, 2048) * 1024).astype(np.int32)
x = jax.numpy.asarray(x)
print(x)

print(x > 100)
print(jnp.sum((x > 100).astype(jax.numpy.int32), -1, keepdims=True))
print(jnp.sum((x > 100).astype(jax.numpy.bfloat16), -1, keepdims=True))

```

Outputs:
```
[[383 973 749 ... 740  69 724]
 [556  83 469 ... 951 339 476]]
[[ True  True  True ...  True False  True]
 [ True False  True ...  True  True  True]]
[[1828]
 [1837]]
[[1824]
 [1840]]
```

Note the difference in the sum!

**Comment by jakevdp** [comment]
You closed this so you've probably figured it out, but in case others stumble across this: the reason for this is that bfloat16 is incapable of representing the correct output in this case. For example:
```python
>>> from ml_dtypes import bfloat16, finfo
>>> bfloat16(1828)
1824
```
Using [`numpy.nextafter`](https://numpy.org/doc/stable/reference/generated/numpy.nextafter.html), we can see that the next representable bfloat16 value after 1824 is 1832:
```python
>>> import numpy as np
>>> np.nextafter(bfloat16(1824), bfloat16(2000))
1832
```
You should always keep this in mind when working with `bfloat16` values: they encode only an 8-bit significand, so computational results will only be accurate to about 1 part in 256. This is by design: `bfloat16` was specifically created for situations where numerical accuracy can be discarded in favor of more efficient computation.
